### PR TITLE
Fix memory leak in _LRScheduler.step()

### LINKED ISSUE #85410: _LRScheduler memory leak
### 🐛 Describe the bug

There are reference cycles in `:meth:_LRScheduler.step`. To reproduce
```python
import gc
import torch
from torch.nn import Linear
from torch.optim import Adam
from torch.optim.lr_scheduler import MultiStepLR


def main():
    t = torch.randn(100, 100)
    model = Linear(100, 100)

    optimizer = Adam(params=model.parameters())
    scheduler = MultiStepLR(optimizer=optimizer, milestones=[1, 2, 3])
    t = model(t)
    t.sum().backward()
    optimizer.step()
    scheduler.step()


if __name__ == '__main__':
    gc.enable()
    gc.set_debug(gc.DEBUG_SAVEALL)
    main()
    garbage = gc.collect()

    try:
        assert not garbage, "Memory leak is occurred!"
    except AssertionError:
        import objgraph
        objgraph.show_backrefs(gc.garbage, filename='finalizers.png')
        raise
```
Please see the generated graph showing the cycle references.
I have to override the `step` on my end by moving this context manager https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L146-L156 outside the method.
I hope you will fix it soon.

### Versions

$ python collect_env.py
Collecting environment information...
PyTorch version: 1.11.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.9.10 | packaged by conda-forge | (main, Feb  1 2022, 21:24:37)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-46-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] gpytorch==1.6.0
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] numpydoc==1.2
[pip3] pytorch-lightning==1.5.4
[pip3] torch==1.11.0
[pip3] torch-cluster==1.6.0
[pip3] torch-geometric==2.0.3
[pip3] torch-scatter==2.0.9
[pip3] torch-sparse==0.6.13
[pip3] torch-spline-conv==1.2.1
[pip3] torchmetrics==0.6.2
[conda] gpytorch                  1.6.0                    pypi_0    pypi
[conda] numpy                     1.21.5                   pypi_0    pypi
[conda] numpydoc                  1.2                      pypi_0    pypi
[conda] pytorch-lightning         1.5.4                    pypi_0    pypi
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torch-cluster             1.6.0                    pypi_0    pypi
[conda] torch-geometric           2.0.3                    pypi_0    pypi
[conda] torch-scatter             2.0.9                    pypi_0    pypi
[conda] torch-sparse              0.6.13                   pypi_0    pypi
[conda] torch-spline-conv         1.2.1                    pypi_0    pypi
[conda] torchmetrics              0.6.2                    pypi_0    pypi

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by KinglittleQ**
Fixes #85410

This diff removed the cyclic references in `_LRScheduler.step()`.

### CODE REVIEW

**Comment by albanD** [comment]
The change sounds ok but the CI failures are real right?

**Comment by KinglittleQ** [comment]
Thanks for your comments. I will fix the CI.

**Comment by KinglittleQ** [comment]
Thanks for your comments. I will fix the CI.

**Comment by KinglittleQ** [comment]
@albanD There are some updates to try to fix the tests. Could you please take another look?

**Comment by albanD** [comment]
Thanks!
(GitHub is in a bad state right now, so CI might take a bit of time to trigger properly).

File: test/test_optim.py [hunk 1]
```diff
@@ -1139,6 +1139,21 @@ def test_no_cyclic_references(self):
...
+            optim = SGD(params=[param], lr=0.5)
+            scheduler = LambdaLR(optim, lambda epoch: 1.0)
+            param.sum().backward()
+            optim.step()
+            scheduler.step()
+
+        import gc
+        gc.collect()
+        run()
+        garbage = gc.collect()
```
Comment (albanD) [inline test/test_optim.py:1154]: You don't want the collect here I think?
THe whole point here is not to create ref cycles right?

Comment (KinglittleQ) [reply test/test_optim.py:1154]: Yes, if there are any ref cycles created in `run()`, the second `gc.collect()` would return non-zero.

**Comment by KinglittleQ** [comment]
@albanD The CI needs to be triggered by a maintainer since I am a first-time contributor. Could you take a look?

File: test/test_optim.py (hunk 1 above)
Comment (albanD) [reply test/test_optim.py:1232]: My bad, I was still under the idea of using weakref and didn't read the update carefully. Sounds good!

**Review by albanD — APPROVED** [review]
Looks good thanks

**Comment by albanD** [comment]
For more information on the flaky test, see https://github.com/pytorch/pytorch/issues/86413

**Comment by albanD** [comment]
@KinglittleQ my best guess of the flakyness here is that if any other object happens to be ready for collection when you run gc.collect(), it will fail the test even though we didn't create a cycle in the function.

Could you update the test to use weakref again so that we make sure that: while the gc is disabled, the weakef remains alive after the function AND after running the gc manually the weakref is gone.

**Comment by KinglittleQ** [comment]
@albanD The test has been updated now.

File: test/test_optim.py [hunk 2]
```diff
@@ -1217,6 +1217,28 @@ def test_no_cyclic_references(self):
...
+
+            return weakref.ref(scheduler)
+
+        # To ensure that there are no reference cycles in scheduler,
+        # we need to turn off the garbage collector. Since gc will
+        # automatically collect unreachable objects.
+        gc.disable()
+        ref = run()
+        assert ref() is None
+        gc.enable()  # restore
```
Comment (KinglittleQ) [inline test/test_optim.py:1240]: The `scheduler` object should be freed after `run()` if there are no ref cycles. 

This test ensures that the weakref will be unavailable even if the gc is disabled.  (gc would collect objects with ref cycles)

**Review by albanD — APPROVED** [review]
Perfect! Thanks!
